icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+    //     ICICLE_LOG_DEBUG << "temp_elements[" << i << "]: " << temp_elements[i];
+    // }
+
+    S* twiddles = s_ntt_domain<S>->get_twiddles();
+
+    // NTT/INTT
+    int ntt_step = 0;
+    if (dit) {
+      ICICLE_LOG_DEBUG << "DIT";
+      for (int batch = 0; batch < config.batch_size; ++batch) {
+        E* current_temp_elements = temp_elements + batch * size;
+        for (int len = 2; len <= size; len <<= 1) {
+          // ICICLE_LOG_DEBUG << "ntt_step: " << ntt_step++;
+          int half_len = len / 2;
+          int step = size / len;
+          int tw_idx = 0;
+          for (int i = 0; i < size; i += len) {
+            for (int j = 0; j < half_len; ++j) {
+              tw_idx = (dir == NTTDir::kForward) ? j * step : size - j * step;
+              E u = current_temp_elements[i + j];
+              E v = current_temp_elements[i + j + half_len] * twiddles[tw_idx];
+              // ICICLE_LOG_DEBUG << "tw_idx=" << tw_idx;
+              // ICICLE_LOG_DEBUG << "current_temp_elements[" << i + j << "] <-- " << current_temp_elements[i + j] << "
+              // + " << current_temp_elements[i + j + half_len] << "*" << twiddles[tw_idx]; ICICLE_LOG_DEBUG <<
+              // "current_temp_elements[" << i + j + half_len << "] <-- " << current_temp_elements[i + j] << " - " <<
+              // current_temp_elements[i + j + half_len] << "*" << twiddles[tw_idx];
+              current_temp_elements[i + j] = u + v;
+              current_temp_elements[i + j + half_len] = u - v;
+              // ICICLE_LOG_DEBUG << i + j << " <-- " << i + j << " + " << i + j + half_len << "*" << tw_idx;
+              // ICICLE_LOG_DEBUG << i + j + half_len << " <-- " << i + j << " - " << i + j + half_len << "*" << tw_idx;
+            }
+          }
+        }
+      }
+    } else { // dif
+      ICICLE_LOG_DEBUG << "DIF";
+      for (int batch = 0; batch < config.batch_size; ++batch) {
+        E* current_temp_elements = temp_elements + batch * size;
+        for (int len = size; len >= 2; len >>= 1) {
+          // ICICLE_LOG_DEBUG << "ntt_step: " << ntt_step++;
+          int half_len = len / 2;
+          int step = size / len;
+          int tw_idx = 0;
+          for (int i = 0; i < size; i += len) {
+            for (int j = 0; j < half_len; ++j) {
+              tw_idx = (dir == NTTDir::kForward) ? j * step : size - j * step;
+              E u = current_temp_elements[i + j];
+              E v = current_temp_elements[i + j + half_len];
+              current_temp_elements[i + j] = u + v;
+              current_temp_elements[i + j + half_len] = (u - v) * twiddles[tw_idx];
+            }
+          }
+        }
+      }
+    }
+
+    if (dir == NTTDir::kInverse) {
+      // Normalize results
+      S inv_size = S::inv_log_size(logn);
+      for (int i = 0; i < total_size; ++i) {
+        temp_elements[i] = temp_elements[i] * inv_size;
+      }
+    }
+
+    if (config.columns_batch) {
+      // Distribute output into columns if columns_batch is set
```

Comment:
since you have it twice I would make it a function named transpose, assuming that's what it is doing

Reply:
fixed